### ISSUE TITLE
Download Rate Limiting & Concurrent Downloads

### ISSUE THREAD

**Issue by romen-h** [Kalanyr/gogrepo#89]
I was wondering if it would be possible to add arguments to control the following behaviour:

- ~~Download rate limit~~  
~~Something like `-ratelimit x` which would ensure the download rate of any individual worker does not exceed x MB/s.~~

- Concurrent downloads (workers)
Something like `-workers x` which would ensure x workers are spawned to do the downloads in parallel.

These two arguments would allow me to use this script like a scheduled task or service that does not consume all of my bandwidth in the background. Downloading one file at a time also helps with resuming if the script is terminated since there will be only one partial file that needs to be re-downloaded instead of 4 partial files from cancelling.

**Comment by Kalanyr** [comment]
Unfortunately no, I can expose the max number of simultaneous downloads (
which get 1 worker each ) but the module used doesn't allow rate limiting
and adding the capacity to the requests module is out of scope.

**Comment by romen-h** [comment]
Thanks for responding.

I will try to find some external way to rate limit just this script, maybe with domain based QoS or something.

I'm still interested in the argument to set the number of workers, but even that might not be necessary if download resuming works. Downloading one file at a time would mean only one part has to be restarted and more parts may have downloaded successfully before an interruption.

**Comment by romen-h** [comment]
Also if you point me to the lines where the HTTP requests for the "chunks" are happening then I might be able to implement this and PR it if it works.

Apparently it's easy to implement rate limiting with the python requests module; You just put a sleep between requests for HTTP chunks so that your overall rate will not exceed `chunk_size / sleep_time`.

**Comment by Kalanyr** [comment]
If you're okay with that I can implement it but it's a very coarse rate
limit, so you should expect to go over your set limit, significantly so, at
times.
